### DISCUSSION TITLE
Proposals: Release 5.X

### DISCUSSION THREAD

**Discussion by jackmatt2** [origin-energy/java-snapshot-testing#144] [General]
Proposal 1: Make in built Serilizers more closely assign with jest format: 
**BREAKING**
Refactor: No
Regenerate Snapshots: Yes

```
exports[`will fail every time 1`] = `
{
  "createdAt": "2018-05-19T23:36:09.816Z",
  "id": 3,
  "name": "LeBron James"
}
`;
```

Resources:

Write Format
https://github.com/facebook/jest/blob/main/packages/jest-snapshot/src/utils.ts#L207-L214
https://github.com/facebook/jest/blob/main/packages/jest-snapshot/src/utils.ts#L217-L220

Handle New Lines
https://github.com/facebook/jest/blob/main/packages/jest-snapshot/src/utils.ts#L201

Escaping
https://github.com/facebook/jest/blob/main/packages/jest-snapshot/src/utils.ts#L189

Parse from filesystem (Use Nashhorn?)
https://github.com/facebook/jest/blob/main/packages/jest-snapshot/src/utils.ts#L113

----------

Proposal 2: Remove flakey parameter injection.  It doesn't work for parameterised tests.
Injecting `expect` as a method parameter does not work in many cases.  Instance level injection work in all cases.
**BREAKING**
Refactor: Yes
Regenerate Snapshots: No

----------

Proposal 3: Respect Junit5 `@DisplayName` as an alias for `@SnapshotName`
**BREAKING**
Refactor: No
Regenerate Snapshots: Yes

**Comment by grimsa** [comment]
> Proposal 1: Make in built Serilizers more closely assign with jest format

If I understand correctly, this would affect the `JacksonSnapshotSerializer`? What are the arguments in favor for this change?
One argument against is that the current format that stores valid JSON allows:
- copy-pasting JSON from snapshot to some other target (e.g. into Postman request body)
- comparing JSON in snapshot with JSON from some different source (e.g. application logs)

My proposal would be to maybe add a new `JestLikeSnapshotSerializer`, and not change the existing ones. This would also make it into a non-breaking change.

**Reply by jackmatt2** [reply]
The format was copied verbatim from the Jest docs.  I have updated the snapshot body (above) now to reflect a more realistic example.  You'll still be able to copy/paste into postman etc. as you can today if you're using the JacksonSerializer.

The main motivation for this change is:
1.  The general template is familiar to JS & Full Stack developers already
```
exports[`foobar 1`] = `

`;
```
2. Custom Serializers are currently a little too flexible.  Spacing issues at the end of the snapshot are causing failures for people and it's difficult for them to understand why. The above consistent template would eliminate those issues.
3. The ToStringSerializer currently appends square brackets around the snapshot.
```
exports[`foobar 1`] = [

]
```
This is a relic of the original Jackson implementation which makes it look too JSON like IMO.  There is a proposal for a YAML Serializer which makes this look quite weird if we follow the same convention.
```
exports[`foobar 1`] = [
name: "Joe",
accounts: 
  - Account 1
  - Account 2
]
```

Updating would be as simple as passing the `-DupdateSnapshot` (I'm also thinking of moving this flag to snapshot.properties cause people are having trouble passing it through to the actual test runner). I'll most probably version the snapshot serializers in the future so what you're saying will become much simpler to do.  I could also backport the original serializers to avoid needing to update the snapshots.

**Reply by grimsa** [reply]
Ah, so this is more about the start/end of the snapshot, rather than the content part itself (about which I was concerned). In that case it would be no problem for us.
Thanks for clarifying.
